Add support for offline planned tensor allocations

jenselofsson opened the pull request:
Add support for offline planning in Tensorflow Lite Micro, and accompanying testcases.

The idea behind the offline memory planner can be found in the following RFC:
https://docs.google.com/document/d/16aTSHL5wxsq99t6adVbBz1U3K8Y5tBDAvs16iroZDEU

It is implemented according to alternative 1 described in section 3.1.

freddan80 committed:
Add support for offline planned tensor allocations

By adding metadata to the model, it is possible to set arena offset
for each tensor.

Change-Id: Idd646c00a6e34e0c2603896d748cd5680a57f015

jenselofsson committed:
OfflinePlanner: Swap offsets in TestAllocationForModelsWithBranches

The offline planner sorts the tensors in reverse order, so the testcase
have to be updated accordingly.

Change-Id: Ic3a1193489d6ad5f592db1c9a289b01083ad9c62

jenselofsson committed:
Testcases for offline planned tensors

Added testcases. Expanded ModelBuilder for have
metadata support.

Change-Id: I7e8bf3b3537d126086aef52bb3d6fe572aa8e7a0

freddan80 committed:
Add helper functions in micro allocator

One function to check metadata correctness and another to print
model data.

Change-Id: I2500fbbac25b376d068e3d9a1d190249da461eef

jenselofsson committed:
Fix compile error when building for ARM Cortex-M4 etc.

freddan80 commented:
"MacOS Python2 and CC" seems unrelated to this change.

wangtz commented on tensorflow/lite/micro/micro_allocator.cc:
```diff
@@ -241,9 +389,17 @@ TfLiteStatus CreatePlan(ErrorReporter* error_reporter, MemoryPlanner* planner,
+                               current->first_created, current->last_used));
+      } else {
+        TF_LITE_ENSURE_STATUS(
+            (static_cast<GreedyMemoryPlanner*>(planner))
```
I don't mind if you change the CreatePlan API to take a GreedyMemoryPlanner to make the requirement explicit.

freddan80 replied:
done

wangtz commented on tensorflow/lite/micro/test_helpers.cc:
```diff
@@ -142,13 +156,34 @@ ModelBuilder::Node ModelBuilder::AddNode(
+                             1 + ModelBuilder::nbr_of_metadata_buffers_);
+
+  metadata_buffers_[nbr_of_metadata_buffers_] = tflite::CreateBuffer(*builder_,
+                                                                       builder_->CreateVector((uint8_t*)metadata_buffer_data,
```
You might want to reformat this code.

wangtz commented on tensorflow/lite/micro/test_helpers.cc:
```diff
@@ -142,13 +156,34 @@ ModelBuilder::Node ModelBuilder::AddNode(
-  const flatbuffers::Offset<Buffer> buffers[kBufferSize] = {
-      tflite::CreateBuffer(*builder_)};
+  size_t kBufferSize = 1 + ModelBuilder::nbr_of_metadata_buffers_;
+  flatbuffers::Offset<Buffer> buffers[kBufferSize];
```
This is dynamic sized array. Not sure if it's allowed on all platforms. 

Alternatively you might also consider other approaches:
- create a large enough list of size kMaxMetadataBuffers?
- append the first two elements to metadata_buffers_ in the constructor.

wangtz commented on tensorflow/lite/micro/test_helpers.cc:
```diff
@@ -159,12 +194,26 @@ const Model* ModelBuilder::BuildModel(
+        builder_->CreateVector(operator_codes_, next_operator_code_id_),
+        builder_->CreateVector(subgraphs, subgraphs_size),
+        builder_->CreateString("teset_model"),
+        builder_->CreateVector(buffers, kBufferSize),
```
k prefix means that it's a const   plz consider a better naming otherwise.

wangtz commented on tensorflow/lite/micro/test_helpers.cc:
```diff
@@ -243,6 +292,35 @@ const Model* BuildSimpleModelWithBranch() {
 
+const Model* BuildModelWithOfflinePlanning(int number_of_tensors,
+    const int32_t* metadata_buffer,
+    std::vector<NodeConnection> node_conn) {
```
We're aiming at making the tests run-able on target device. If std::vector is pulling in a large dependency it might not run in the current build system (we had -nostdlib) or it might be too large to fit in. 

Wondering if there is a way to avoid std::vector?

jenselofsson replied:
I'll see if I can replace it with an array.

wangtz reviewed (commented):
Sry for taking so long! Looks good in general.

gbaned commented:
@jenselofsson Can you please resolve conflicts? Thanks!

freddan80 commented on tensorflow/lite/micro/memory_planner/greedy_memory_planner.h:
```diff
@@ -110,16 +118,23 @@ class GreedyMemoryPlanner : public MemoryPlanner {
+  int* buffer_ids_sorted_;
   ListEntry* buffers_sorted_by_offset_;
   int next_free_entry_;
+  int first_entry_index_;
```
In previous implementation the smallest buffer offset was always located at first in the list (buffers_sorted_by_offset_[0]). This may not be the case when offline planned tensors are present. Therefore, first_entry_index_, keeps track of what index in the list has the smallest buffer offset.

freddan80 commented on tensorflow/lite/micro/micro_allocator.cc:
```diff
@@ -159,6 +257,11 @@ TfLiteStatus AllocationInfoBuilder::AddTensors(const SubGraph* subgraph,
     current->needs_allocating = (runtime_tensors[i].data.raw == nullptr) &&
                                 (!subgraph->tensors()->Get(i)->is_variable());
+    if (offline_offsets) {
+      current->offline_offset = offline_offsets[i];
```
Yes, I'll add that comment

jenselofsson commented:
@gbaned Taking a look!

freddan80 commented on tensorflow/lite/micro/micro_allocator.cc:
```diff
@@ -216,6 +319,51 @@ TfLiteStatus AllocationInfoBuilder::AddTensors(const SubGraph* subgraph,
+  if (model->metadata()) {
+    for (int i = 0; i < model->metadata()->size(); ++i) {
+      auto metadata = model->metadata()->Get(i);
+      if (strncmp(metadata->name()->c_str(), "OfflineMemoryAllocation",
```
Good point

freddan80 commented on tensorflow/lite/micro/micro_allocator.cc:
```diff
@@ -95,6 +96,94 @@ TfLiteStatus AllocateVariables(
   return kTfLiteOk;
 }
 
+// Helper function to print model flatbuffer data. This function is not called
```
Good idea to move it to tensorflow/lite/micro/micro_optional_debug_tools.cc. Is it ok if I hook it into ther person detection test? To get a real model...

wangtz replied:
For sure.

freddan80 committed:
Fix review comments, 24/4

freddan80 committed:
Merge remote-tracking branch 'origin/upstream/master' into offline_memory_planner

jenselofsson committed:
Address reviewer comments.

jenselofsson commented:
@wangtz Addressed the review comments!

wangtz reviewed (dismissed):
Thanks!

freddan80 commented:
Let us know if there's anything more we should do. Otherwise, I think it's ready to pull.

gbaned commented:
@jenselofsson Can you please check @wangtz comments and keep us posted. Thanks!

jenselofsson commented:
@gbaned Done!

freddan80 commented:
Ready to merge

gbaned commented:
@jenselofsson Can you please resolve conflicts? Thanks!

jenselofsson committed:
Merge remote-tracking branch 'upstream/master' into offline_memory_planner

jenselofsson committed:
Realign AllocationInfo struct.

After adding offline_offset, sizeof(AllocationInfo) = 40, which caused
hello_world_test to crash. After realigning it's back to its original
size (32).

jenselofsson commented:
@petewarden Ping for review

jenselofsson commented:
The Windows Bazel GPU build is failing with this error message:
ERROR: T:/src/github/tensorflow/tensorflow/core/kernels/BUILD:362:1: C++ compilation of rule '//tensorflow/core/kernels:extract_volume_patches_op_gpu' failed (Exit 1)

which is unrelated to Tensorflow Lite Micro.

jenselofsson commented on tensorflow/lite/micro/examples/person_detection/person_detection_test.cc:
```diff
@@ -22,6 +22,7 @@ limitations under the License.
 #include "tensorflow/lite/micro/micro_error_reporter.h"
 #include "tensorflow/lite/micro/micro_interpreter.h"
 #include "tensorflow/lite/micro/micro_mutable_op_resolver.h"
+#include "tensorflow/lite/micro/micro_optional_debug_tools.h"
```
Debug code, should be removed

jenselofsson commented on tensorflow/lite/micro/examples/person_detection/person_detection_test.cc:
```diff
@@ -46,6 +47,7 @@ TF_LITE_MICRO_TEST(TestInvoke) {
                          "to supported version %d.\n",
                          model->version(), TFLITE_SCHEMA_VERSION);
   }
+  PrintModelData(model, error_reporter);
```
Debug code, remove.

jenselofsson commented:
@gbaned Resolved!

jenselofsson committed:
Merge remote-tracking branch 'upstream/master' into offline_memory_planner

jenselofsson commented:
Ready to pull

jenselofsson commented:
@wangtz @petewarden Ping for review!

freddan80 commented:
@gbaned ready to pull

rthadur reviewed (changes requested):
`/tensorflow/lite/micro/micro_allocator.cc:117:18: error: declaration shadows a local variable
        for (int i = 0; i < nbr_offline_offsets; ++i) {
                 ^/tensorflow/lite/micro/micro_allocator.cc:93:14: note: previous declaration is here
    for (int i = 0; i < model->metadata()->size(); ++i) {
             ^/tensorflow/lite/micro/micro_allocator.cc:93:23: error: comparison of integers of different signs: 'int' and 'flatbuffers::uoffset_t' (aka 'unsigned int') 
    for (int i = 0; i < model->metadata()->size(); ++i) {
                    ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~
/tensorflow/lite/micro/micro_allocator.cc:304:23: error: comparison of integers of different signs: 'int' and 'flatbuffers::uoffset_t' (aka 'unsigned int')

    for (int i = 0; i < model->metadata()->size(); ++i) {
                    ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~
/tensorflow/lite/micro/micro_allocator.cc:316:27: error: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'const int32_t' (aka 'const int') 
        if (tensor_count_ != nbr_tensors) {
            ~~~~~~~~~~~~~ ^  ~~~~~~~~~~~
4 errors generated.
/tensorflow/lite/micro/micro_optional_debug_tools.cc:127:11: error: unused variable 'quantization' 
    auto* quantization = flatbuffer_tensor.quantization();
          ^
third_party/tensorflow/lite/micro/micro_optional_debug_tools.cc:125:21: error: comparison of integers of different signs: 'int' and 'flatbuffers::uoffset_t' (aka 'unsigned int') 
  for (int i = 0; i < tensors->size(); ++i) {
                  ~ ^ ~~~~~~~~~~~~~~~
2 errors generated.

third_party/tensorflow/lite/micro/test_helpers.cc:311:14: error: variable length array used 
  int tensors[number_of_tensors];`

rthadur commented:
@jenselofsson can you please check above errors ?

freddan80 commented:
Hi @rthadur ! Thanks for catching this. All valid, we will fix.

When we ran regression using:

`make -f tensorflow/lite/micro/tools/make/Makefile test` 

and running the script:

`./tensorflow/lite/micro/tools/ci_build/test_all.sh`

these errors weren't caught. Will you let us know how you run your tests so we can avoid these issues in the future?

Cheers!

freddan80 committed:
Fixed review comments 27/5

jenselofsson committed:
Merge remote-tracking branch 'upstream/master' into offline_memory_planner

rthadur commented:
@freddan80 these are whole lot of tests which run internally,which iam not aware , @petewarden could you please help with @freddan80 ask , 
Can you please resolve conflicts as well.

jenselofsson commented:
@rthadur Conflicts solved!

jenselofsson commented:
@rthadur Fixed build issues.

jenselofsson committed:
Merge remote-tracking branch 'upstream/master' into offline_memory_planner

jenselofsson commented:
Fixed merge conflicts.

jenselofsson commented:
Ready to merge

rthadur commented:
@jenselofsson sorry for the delay it got stuck internally , please check below errors 

`/tensorflow/lite/micro/micro_allocator.cc:102:14: error: unused function 'CheckOfflinePlannedOffsets' [-Werror,-Wunused-function]
TfLiteStatus CheckOfflinePlannedOffsets(const Model* model,
             ^
1 error generated.
Broken by missing target 
/tensorflow/lite/micro/memory_planner:greedy_memory_planner
third_party/tensorflow/lite/micro/memory_planner/greedy_memory_planner.cc:194:15: error: declaration shadows a local variable [-Werror,-Wshadow]
    const int buffer_id = buffer_ids_sorted_[i];
              ^
/tensorflow/lite/micro/memory_planner/greedy_memory_planner.cc:184:7: note: previous declaration is here
  int buffer_id = buffer_ids_sorted_[0];
      ^
1 error generated.`

I guess the tests which generated above error is 

/tensorflow/lite/experimental/microfrontend/lib/BUILD?q=fft_test_binary

jenselofsson committed:
Fix compile errors.

jenselofsson commented:
@rthadur Fixed the compile issues

rthadur commented on tensorflow/lite/micro/micro_allocator.cc:
```diff
@@ -91,6 +94,70 @@ TfLiteStatus AllocateVariables(
+TfLiteStatus CheckOfflinePlannedOffsets(const Model* model,
+                                        ErrorReporter* error_reporter) {
+  // Suppress compile warning for unused function
+  (void)CheckOfflinePlannedOffsets;
```
`tensorflow/lite/micro/micro_allocator.cc:102:14: error: function 'CheckOfflinePlannedOffsets' is not needed and will not be emitted [-Werror,-Wunneeded-internal-declaration]
TfLiteStatus CheckOfflinePlannedOffsets(const Model* model,`

jenselofsson replied:
From what I could see when I tested it, that error showed up when using clang.
I solved it like this, which might not be optimal. What do you think about it?
https://github.com/tensorflow/tensorflow/pull/38121/commits/bacc5e5927360ad119cdb9967d53df4c95c53fd4#diff-6dace628e5372fb3c3fbb76e0fd9d37aR97

rthadur reviewed (changes requested):
@jenselofsson  can you please check below error

gbaned commented:
@jenselofsson Can you please check @rthadur comments and resolve conflicts?. Thanks!

jenselofsson committed:
Fix build issues.

jenselofsson committed:
Merge remote-tracking branch 'upstream/master' into offline_memory_planner

rthadur commented:
@jenselofsson sorry for the delay. , this changes have been failing internally , here is the error log 

`tensorflow/lite/micro/micro_allocator_test.cc:291:26: error: no matching constructor for initialization of 'tflite::MicroAllocator'
 tflite::MicroAllocator allocator(&context, model, arena, arena_size,
                        ^         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tensorflow/lite/micro/micro_allocator.h:143:3: note: candidate constructor not viable: requires 4 arguments, but 5 were provided
 MicroAllocator(TfLiteContext* context, const Model* model,
 ^
/tensorflow/lite/micro/micro_allocator.h:88:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 5 were provided
class MicroAllocator {
     ^
/tensorflow/lite/micro/micro_allocator_test.cc:340:26: error: no matching constructor for initialization of 'tflite::MicroAllocator'
 tflite::MicroAllocator allocator(&context, model, arena, arena_size,
                        ^         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tensorflow/lite/micro/micro_allocator.h:143:3: note: candidate constructor not viable: requires 4 arguments, but 5 were provided
 MicroAllocator(TfLiteContext* context, const Model* model,
 ^
/tensorflow/lite/micro/micro_allocator.h:88:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 5 were provided
class MicroAllocator {
     ^
/tensorflow/lite/micro/micro_allocator_test.cc:385:26: error: no matching constructor for initialization of 'tflite::MicroAllocator'
 tflite::MicroAllocator allocator(&context, model, arena, arena_size,
                        ^         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tensorflow/lite/micro/micro_allocator.h:143:3: note: candidate constructor not viable: requires 4 arguments, but 5 were provided
 MicroAllocator(TfLiteContext* context, const Model* model,
 ^
/tensorflow/lite/micro/micro_allocator.h:88:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 5 were provided
class MicroAllocator {
     ^
/tensorflow/lite/micro/micro_allocator_test.cc:433:26: error: no matching constructor for initialization of 'tflite::MicroAllocator'
 tflite::MicroAllocator allocator(&context, model, arena, arena_size,
                        ^         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tensorflow/lite/micro/micro_allocator.h:143:3: note: candidate constructor not viable: requires 4 arguments, but 5 were provided
 MicroAllocator(TfLiteContext* context, const Model* model,
`

jenselofsson committed:
Merge remote-tracking branch 'upstream/master' into offline_memory_planner

jenselofsson committed:
Fix compile errors by using the new MicroAllocator::Create()

rthadur committed:
Merge branch 'master' into offline_memory_planner

rthadur commented:
@jenselofsson here is the internally error , can you please check 

`tensorflow/lite/micro/micro_allocator.cc:895:72: error: too few arguments to function call, expected 3, have 2
    TF_LITE_ENSURE_STATUS(builder.AddTensors(subgraph, context->tensors));
                          ~~~~~~~~~~~~~~~~~~                           ^
./tensorflow/lite/c/common.h:198:29: note: expanded from macro 'TF_LITE_ENSURE_STATUS'
    const TfLiteStatus s = (a);  \
                            ^
/tensorflow/lite/micro/micro_allocator.cc:225:37: note: 'AddTensors' declared here
TfLiteStatus AllocationInfoBuilder::AddTensors(const SubGraph* subgraph,
                                    ^
1 error generated`

jenselofsson committed:
Adapt to changes in micro_allocator.

jenselofsson commented:
@rthadur Fixed build issues

jenselofsson committed:
Merge remote-tracking branch 'upstream/master' into offline_memory_planner
